libsolidity/formal/EncodingContext.cpp

```diff
@@ -28,14 +28,57 @@ EncodingContext::EncodingContext(SolverInterface& _solver):
 	m_solver(_solver),
 	m_thisAddress(make_unique<SymbolicAddressVariable>("this", m_solver))
 {
+	auto sort = make_shared<smt::ArraySort>(
+		make_shared<smt::Sort>(smt::Kind::Int),
+		make_shared<smt::Sort>(smt::Kind::Int)
+	);
+	m_balances = make_unique<SymbolicVariable>(sort, "balances", m_solver);
 }
 
 void EncodingContext::reset()
 {
 	m_thisAddress->increaseIndex();
+	m_balances->increaseIndex();
 }
 
 smt::Expression EncodingContext::thisAddress()
 {
 	return m_thisAddress->currentValue();
 }
+
+smt::Expression EncodingContext::balance()
+{
+	return balance(m_thisAddress->currentValue());
+}
+
+smt::Expression EncodingContext::balance(smt::Expression _address)
+{
+	return smt::Expression::select(m_balances->currentValue(), move(_address));
+}
+
+void EncodingContext::transfer(smt::Expression _to, smt::Expression _value)
+{
+	unsigned indexBefore = m_balances->index();
+	addBalance(m_thisAddress->currentValue(), 0 - _value);
+	addBalance(_to, move(_value));
+	unsigned indexAfter = m_balances->index();
+	solAssert(indexAfter > indexBefore, "");
+	m_balances->increaseIndex();
+	auto newBalances = smt::Expression::ite(
```

Comment:
Can you explain this part?

Reply:
If one is transferring to `address(this)` the transfer is ignored.

Reply:
Wrote it mainly because without the `ite`, the `insufficient balance` SMT would always give UNSAT due to `0 - _value` underflowing and `balance[_to] + _value` overflowing.

Reply:
Ah, I see. Wouldn't it be better to use a generic function with `transfer(from, to, value)`?

Reply:
In any case, could you add a comment?

Reply:
Also, does `address(this).transfer(address(this).balance + 1)` cause a revert at runtime?

Reply:
Yes, it reverts

Reply:
I'll add the generic function